File: src/apple/Packages/MapleUI/Sources/MapleUI/Internal/MuiCurvePlotMath.swift
// MuiCurvePlotMath.swift — pure point/curve math for MuiCurvePlot
// (unified-component-catalog.md §2.6). Mirrors the web reference's
// `mui-curve-plot`: points are 0...1 normalized (x right-positive, y
// up-positive — canvas y is flipped since screen y grows down), hit-tested
// by pixel radius, and smoothed with the same "midpoint-quadratic" curve —
// each interior control point bends the curve without ever overshooting
// past it, since the curve's abscissa is a Bernstein blend of three
// non-decreasing x's (prior midpoint <= this point's x <= next midpoint)
// whenever the input points are sorted by x. That's what keeps the curve
// monotone in x by construction, the way a tone curve must be.

import Foundation
import CoreGraphics

enum MuiCurvePlotMath {
    static let hitRadius: CGFloat = 8
    static let nudgeStep = 0.02

    static func clampUnit(_ v: Double) -> Double {
        Swift.max(0, Swift.min(1, v))
    }

    /// Normalized point -> canvas pixel, for a plot of `width`x`height`.
    static func toCanvasPoint(_ p: MuiCurvePoint, width: CGFloat, height: CGFloat) -> CGPoint {
        CGPoint(x: p.x * width, y: (1 - p.y) * height)
    }

    /// Canvas pixel -> normalized point, clamped into `[0, 1]`. The inverse
    /// of `toCanvasPoint`.
    static func fromCanvasPoint(_ point: CGPoint, width: CGFloat, height: CGFloat) -> MuiCurvePoint {
        guard width > 0, height > 0 else { return MuiCurvePoint(x: 0, y: 0) }
        return MuiCurvePoint(
            x: clampUnit(Double(point.x / width)),
            y: clampUnit(1 - Double(point.y / height))
        )
    }

    /// The index of whichever point's canvas position is within
    /// `hitRadius` of `location`, or `nil` if none qualify. First match
    /// wins (matches the web reference's linear scan).
    static func hitTest(_ points: [MuiCurvePoint], location: CGPoint, width: CGFloat, height: CGFloat) -> Int? {
        for (index, p) in points.enumerated() {
            let c = toCanvasPoint(p, width: width, height: height)
            let dx = c.x - location.x
            let dy = c.y - location.y
            if (dx * dx + dy * dy).squareRoot() <= hitRadius { return index }
        }
        return nil
    }

    /// Arrow-key nudge: moves one point by `(dx, dy)` normalized units,
    /// clamped back into `[0, 1]`.
    static func nudged(_ p: MuiCurvePoint, dx: Double, dy: Double) -> MuiCurvePoint {
        MuiCurvePoint(x: clampUnit(p.x + dx), y: clampUnit(p.y + dy))
    }

    /// `points` converted to canvas pixels and sorted by x — the order the
    /// curve is actually drawn in, regardless of the order points were
    /// supplied or dragged into.
    static func sortedCanvasPoints(_ points: [MuiCurvePoint], width: CGFloat, height: CGFloat) -> [CGPoint] {
        points.sorted { $0.x < $1.x }.map { toCanvasPoint($0, width: width, height: height) }
    }

    /// One interior "curve to" step of the midpoint-quadratic path:
    /// `control` bends the curve, `end` is the midpoint it lands on (the
    /// final point's own true position is a plain trailing line, not part
    /// of this list — see `MuiCurvePlot`'s `Path` construction).
    struct Segment: Equatable {
        let control: CGPoint
        let end: CGPoint
    }

    /// Builds the interior midpoint-quadratic segments for an already
    /// sorted-by-x canvas point list (`sortedCanvasPoints`'s output). The
    /// drawn path starts at `points[0]`, runs each of these as a quadratic
    /// curve, then finishes with a plain line to `points.last`.
    static func segments(forSortedCanvasPoints points: [CGPoint]) -> [Segment] {
        guard points.count >= 2 else { return [] }
        guard points.count > 2 else { return [] }
        var result: [Segment] = []
        for i in 1..<(points.count - 1) {
            let mid = CGPoint(x: (points[i].x + points[i + 1].x) / 2, y: (points[i].y + points[i + 1].y) / 2)
            result.append(Segment(control: points[i], end: mid))
        }
        return result
    }

    /// A point at parameter `t` (`0...1`) along the quadratic Bezier
    /// `start -> control -> end`. Exposed for monotonicity testing — the
    /// view itself never needs to sample a curve it hands straight to
    /// `Path.addQuadCurve`.
    static func quadraticBezierPoint(start: CGPoint, control: CGPoint, end: CGPoint, t: Double) -> CGPoint {
        let mt = 1 - t
        let x = mt * mt * Double(start.x) + 2 * mt * t * Double(control.x) + t * t * Double(end.x)
        let y = mt * mt * Double(start.y) + 2 * mt * t * Double(control.y) + t * t * Double(end.y)
        return CGPoint(x: x, y: y)
    }
}

File: src/windows/Maple.WinUI/MapleUI/Molecules/MuiCurvePlotMath.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Maple.UI
{
    /// <summary>One draggable Curve Plot control point, normalized 0..1 on
    /// both axes.</summary>
    public readonly record struct MuiCurvePoint(double X, double Y);

    /// <summary>One drawn segment of a smoothed curve path, in normalized
    /// 0..1 model space (the caller scales to canvas pixels).</summary>
    public abstract record MuiCurveSegment;

    /// <summary>A straight line to <see cref="To"/> — used only for the
    /// final segment into the last (pinned) point.</summary>
    public sealed record MuiCurveLineTo(MuiCurvePoint To) : MuiCurveSegment;

    /// <summary>A quadratic Bezier curve to <see cref="To"/> using
    /// <see cref="Control"/> as its control point.</summary>
    public sealed record MuiCurveQuadTo(MuiCurvePoint Control, MuiCurvePoint To) : MuiCurveSegment;

    /// <summary>
    /// Plain, WinUI-free drag/hit-test/smoothing math behind the Maple.UI
    /// Curve Plot data plot (unified-component-catalog.md §2.6). Same split
    /// as <see cref="MuiSliderMath"/> — linkable into Maple.WinUI.Tests
    /// without a live Window. Ports `mui-curve-plot.component.ts`'s
    /// hit-test/canvas-conversion/nudge math AND its `draw()` method's
    /// midpoint-quadratic path construction exactly — same "always monotone
    /// in x, no overshoot" stand-in the web port uses in place of a full
    /// Fritsch-Carlson monotone spline (deliberately NOT the app's own
    /// 64-segment monotone-cubic <c>ToneCurveMath</c>, which is a different,
    /// richer curve used by the real tone-curve editor — this is the
    /// generic "Curve Plot" primitive, matching the web/Swift ports' own
    /// lighter-weight choice for that primitive).
    /// </summary>
    public static class MuiCurvePlotMath
    {
        public const double HitRadiusPx = 8;
        public const double NudgeStep = 0.02;

        public static double Clamp01(double v) => Math.Max(0, Math.Min(1, v));

        /// <summary>Normalized model point to canvas pixel point — Y is
        /// flipped (model Y grows up, canvas Y grows down).</summary>
        public static MuiCurvePoint ToCanvasPoint(MuiCurvePoint p, double width, double height) =>
            new(p.X * width, (1 - p.Y) * height);

        /// <summary>Canvas pixel position to a clamped normalized model
        /// point. A non-positive width/height reads as 0 on that axis
        /// (no measured canvas yet) rather than dividing by zero.</summary>
        public static MuiCurvePoint FromCanvasPoint(double x, double y, double width, double height) => new(
            Clamp01(width > 0 ? x / width : 0),
            Clamp01(height > 0 ? 1 - y / height : 0));

        /// <summary>Index of the first point within <see cref="HitRadiusPx"/>
        /// of the given canvas position, or null when none is close enough.
        /// <paramref name="points"/> are normalized model points; the
        /// canvas-space comparison happens internally.</summary>
        public static int? HitTest(IReadOnlyList<MuiCurvePoint> points, double canvasX, double canvasY, double width, double height)
        {
            for (var i = 0; i < points.Count; i++)
            {
                var c = ToCanvasPoint(points[i], width, height);
                var dx = c.X - canvasX;
                var dy = c.Y - canvasY;
                if (Math.Sqrt(dx * dx + dy * dy) <= HitRadiusPx) return i;
            }
            return null;
        }

        /// <summary>Arrow-key nudge of one point by <see cref="NudgeStep"/>,
        /// clamped into [0,1] on both axes. An unrecognized key name
        /// returns the point unchanged.</summary>
        public static MuiCurvePoint Nudge(MuiCurvePoint p, string key) => key switch
        {
            "Up" => p with { Y = Clamp01(p.Y + NudgeStep) },
            "Down" => p with { Y = Clamp01(p.Y - NudgeStep) },
            "Right" => p with { X = Clamp01(p.X + NudgeStep) },
            "Left" => p with { X = Clamp01(p.X - NudgeStep) },
            _ => p,
        };

        /// <summary>Builds the same midpoint-quadratic smoothed path
        /// `mui-curve-plot.component.ts`'s `draw()` traces: points sorted by
        /// X, a quadratic segment from each interior point to the midpoint
        /// of it and its successor (using the interior point itself as the
        /// control point), and a final straight line into the last point.
        /// Fewer than 2 points yields an empty path — <paramref name="start"/>
        /// is the single point (or the default point for an empty list) in
        /// that case, since there's nothing to connect it to.</summary>
        public static IReadOnlyList<MuiCurveSegment> BuildSmoothedPath(
            IReadOnlyList<MuiCurvePoint> points, out MuiCurvePoint start)
        {
            var sorted = points.OrderBy(p => p.X).ToList();
            if (sorted.Count < 2)
            {
                start = sorted.Count == 1 ? sorted[0] : default;
                return Array.Empty<MuiCurveSegment>();
            }

            start = sorted[0];
            var segments = new List<MuiCurveSegment>();
            for (var i = 1; i < sorted.Count - 1; i++)
            {
                var mid = new MuiCurvePoint(
                    (sorted[i].X + sorted[i + 1].X) / 2,
                    (sorted[i].Y + sorted[i + 1].Y) / 2);
                segments.Add(new MuiCurveQuadTo(sorted[i], mid));
            }
            segments.Add(new MuiCurveLineTo(sorted[^1]));
            return segments;
        }
    }
}
